### ISSUE TITLE
Matcaffe: output score value is not between [0 and 1]

### ISSUE THREAD

**Issue by 201power** [BVLC/caffe#2515]
I recently started to use matcaffe and it seems the output value is not between [0 and 1] as the python caffe (after softmax). The output score of matcaffe could be minus or larger than one. 

I just used the default googlenet. 

Anyone experience the same issue?

**Comment by ronghanghu** [comment]
This should not have happened, even with the current matlab wrapper (which doesn't support ND-array). There could be some bug. Could you paste here your code to reproduce this problem?

**Comment by 201power** [comment]
Here is the matlab code:

function [] = caffe_classify(inputFolder)

	%% model
	use_gpu = 1;

	model_folder = '/home/lh/Downloads/20150522-googlenet-single-70';
	model_file=[model_folder filesep 'snapshot_iter_163030.caffemodel'];
	
	disp(model_folder);
	model_def_file=[model_folder filesep 'deploy.prototxt'];
	matcaffe_init(use_gpu,model_def_file, model_file);
	% read image mean
	modelmean=caffe('read_mean',[model_folder filesep 'mean.binaryproto']);
	% read labels
	trainType = importdata([model_folder filesep 'labels.txt']);

	%% data
	inputFolder = '/media/test/data_single';
	flist = dir(inputFolder);

	ch = 3;
	ndef = 1;
	predict = [];

	for i=3:numel(flist)
		imgfile = dir([inputFolder filesep flist(i).name filesep '*.png']);
		tic;
		for j=1:numel(imgfile)
			% prepare oversampled input
			% input_data is Height x Width x Channel 
			for k=1:ch
				im(:,:,k) = imread([inputFolder filesep flist(i).name filesep imgfile(j+k-1).name]);
				imgs(:,:,k,1) = prepare_image(im(:,:,k),256,224,modelmean,1);
			end
			% do forward pass to get scores
			scores = [];
			for k=1:size(imgs,3)
				s = caffe('forward',{imgs(:,:,k,1)});
				scores(:,:,k) = s{1};    
			end
			scores = squeeze(scores);
			sepScore{ndef} = scores;
			score(i,:) = mean(scores,2)';
			[~,maxlabel] = max(score(i,:));
			predict(i)=trainType(maxlabel);
			
			ndef = ndef + 1;
		end % end loop image
	end % end 
    
end

% ------------------------------------------------------------------------
function images = prepare_image(img,IMAGE_DIM,CROPPED_DIM,IMAGE_MEAN,nimg)
    % ------------------------------------------------------------------------

    % resize to fixed input size
    % transpose
    for i=1:size(img,3)
        im(:,:,i) = img(:,:,i)';
    end
    im = single(im);
    im = imresize(im, [IMAGE_DIM IMAGE_DIM], 'bilinear');
    % permute from RGB to BGR (IMAGE_MEAN is already BGR)

    im = im - IMAGE_MEAN;

    images = zeros(CROPPED_DIM, CROPPED_DIM, 1, nimg, 'single');
    indices = [0 IMAGE_DIM-CROPPED_DIM] + 1;
    if nimg == 1
        % center crop
        border = fix((IMAGE_DIM-CROPPED_DIM )/2);
        images = im(border:border+CROPPED_DIM-1, border:border+CROPPED_DIM-1, :);
        return;
    end
end

**Comment by ronghanghu** [comment]
I suspect this is more likely to be an usage issue than a bug in Matcaffe. Your network is probably output something other than 'prob', or having multiple output blobs and the first one is not 'prob' produced by a Softmax layer.

You should use BVLC's GoogLeNet model in ./models/bvlc_googlenet/deploy.prototxt` and `./models/bvlc_googlenet/bvlc_googlenet.caffemodel`. You can download it from Caffe's model zoo by running `python ./scripts/download_model_binary.py ./models/bvlc_googlenet/`, and see if you see this problem again.

During initialization, you should see something like 

    ...
    I0527 12:59:11.914986  6326 net.cpp:194] conv1/7x7_s2 does not need backward computation.
    I0527 12:59:11.915010  6326 net.cpp:235] This network produces output prob
    I0527 12:59:11.915289  6326 net.cpp:482] Collecting Learning Rate and Weight Decay.
    ...

outputted to command line stderr to be sure the network is actually having one output blob 'prob'.

I tried several times the following issue.

    addpath ./matlab/caffe

    model = './models/bvlc_googlenet/deploy.prototxt';
    weights = './models/bvlc_googlenet/bvlc_googlenet.caffemodel';
    caffe('set_mode_gpu');
    caffe('init', model, weights, 'test');

    data = rand([224 224 3 10], 'single');
    res = caffe('forward', {data});
    prob = res{1};

    fprintf('prob min: %f, prob max: %f\n', min(prob(:)), max(prob(:)));

and it always gives me prob between 0 and 1. I believe this is not a Matcaffe issue, but a usage issue. Usage issues should be asked at https://groups.google.com/forum/#!forum/caffe-users

**Comment by 201power** [comment]
I found the issue. The network
![image](https://cloud.githubusercontent.com/assets/7299296/7829340/9f163094-03f5-11e5-8694-b6c53d45fe80.png)
